src/v/cluster/cluster_discovery.cc

```diff
@@ -142,8 +202,259 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+          clusterlog.error,
+          "Cluster configuration error: empty_seed_starts_cluster "
+          "mismatch, local: {}, {}: {}",
+          config::node().empty_seed_starts_cluster(),
+          peer_addr,
+          peer_reply.empty_seed_starts_cluster);
+        failure = true;
+    }
+}
+
+} // namespace
+
+cluster_discovery::brokers
+cluster_discovery::get_seed_brokers_for_bootstrap() const {
+    vassert(
+      _cluster_bootstrap_info_fetched,
+      "Cluster bootstrap info not fetched yet");
+
+    // Validate the returned replies and populate the brokers vector
+    bool failed = false;
+    brokers seed_brokers;
+    seed_brokers.reserve(_cluster_bootstrap_info.size());
+    for (const auto& peer : _cluster_bootstrap_info) {
+        if (
+          peer.second.version
+          != features::feature_table::get_latest_logical_version()) {
+            vlog(
+              clusterlog.error,
+              "Cluster setup error: logical version mismatch, local: {}, {}: "
+              "{}",
+              features::feature_table::get_latest_logical_version(),
+              peer.first,
+              peer.second.version);
+            failed = true;
+        }
+        if (peer.second.cluster_uuid.has_value()) {
+            vlog(
+              clusterlog.error,
+              "Cluster setup error: peer is already a member of cluster {}",
+              *peer.second.cluster_uuid);
+            failed = true;
+        }
+        check_configuration_match(failed, peer.first, peer.second);
+        seed_brokers.push_back(peer.second.broker);
+    }
+    if (failed) {
+        throw std::runtime_error(fmt_with_ctx(
+          fmt::format,
+          "Cannot bootstrap a cluster due to seed servers mismatch, check "
+          "the log above for details"));
+    }
+
+    // The other seeds will likely mostly all have -1 as their node_id if their
+    // node_id was not set explicitly. Have the returned seed broker node_id
+    // populated with indices, relying on the fact that all seed broker's have
+    // identical configuration. Note that the order is preserved from
+    // seed_servers to seed_brokers, but the latter is missing one item for
+    // self.
+    const std::vector<config::seed_server>& self_seed_servers
+      = config::node().seed_servers();
+    for (auto [ib, is]
+         = std::tuple{seed_brokers.begin(), self_seed_servers.cbegin()};
+         ib != seed_brokers.cend();
+         ++ib) {
+        while (is != self_seed_servers.cend()
+               && is->addr != ib->rpc_address()) {
+            ++is;
+        }
+        if (ib->id() == model::unassigned_node_id) {
+            vassert(
+              is != self_seed_servers.cend(),
+              "seed_brokers and seed_servers mismatch");
+            const auto idx = boost::numeric_cast<model::node_id::type>(
+              std::distance(self_seed_servers.cbegin(), is));
+            ib->replace_unassigned_node_id(model::node_id{idx});
+        }
+    }
```

Comment:
nit: I'm having a really hard time understanding this code. Does this work instead?
```
(instead of populating seed_brokers above, do it here?)
model::node_id::type idx = 0;
brokers seed_brokers;
for (const auto& ss : self_seed_servers) {
    auto& broker = _cluster_bootstrap_info[ss.addr].broker;
    if (broker.id == model::unassigned_node_id) {
        broker.replace_unassigned_node_id(model::node_id(idx));
    }
    seed_brokers.emplace_back(std::move(broker));
    ++idx;
}
```

Reply:
i would follow andrew suggestion here, seems like a major simplification